### PR TITLE
Implemented all create_ec2.sh functionality using terraform

### LINKED ISSUE #43: Provisioning: validate docker is installed
Connect to the instance we created, and validate that docker is installed

### LINKED ISSUE #47: Infra: Configure the policy, IAM SSH key-pair and ec2 creation in terraform

### PR DESCRIPTION
**Pull request by yerdreic**
- Created policy
- Created IAM role
- Attached the policy to the role
- Created an instance profile, and added the role to that instance profile
- Added configuration to the instance creation so the instance profile was created will be attached to it
- Added secret creation

### CODE REVIEW

File: infra/main.tf [hunk 1]
```diff
@@ -27,10 +27,118 @@ resource "aws_security_group" "allow_ssh" {
...
+  roles      = [aws_iam_role.poc_ec2_role.name]
+  policy_arn = aws_iam_policy.poc_ec2_policy.arn
+}
+
+resource "aws_iam_instance_profile" "poc_ec2_profile" {
+  name = var.profile_name
+  role = aws_iam_role.poc_ec2_role.name
+}
+
+# Generate a private TLS key using ED25519 encryption with a length of 2048 bits
```
Comment (lmilbaum) [inline infra/main.tf]: Is that commented needed?

File: infra/main.tf [hunk 2]
```diff
@@ -23,14 +24,145 @@ resource "aws_security_group" "allow_ssh" {
...
+  policy_arn = aws_iam_policy.poc_ec2_policy.arn
+}
+
+resource "aws_iam_instance_profile" "poc_ec2_profile" {
+  name = var.profile_name
+  role = aws_iam_role.poc_ec2_role.name
+}
+
+
+# resource "tls_private_key" "poc_rsa" {
```
Comment (lmilbaum) [inline infra/main.tf]: Remove the comments if not needed

File: infra/main.tf [hunk 3]
```diff
@@ -23,14 +24,145 @@ resource "aws_security_group" "allow_ssh" {
...
+# content  = tls_private_key.rsa.private_key_pem
+# filename = "tf-key-pair"
+# }
+
 
-  tags = var.tags
+
+
+
+# Generate a private TLS key using RSA
```
Comment (lmilbaum) [inline infra/main.tf]: Same. Comments are used only to explain why the following code was used. Not what it does.

File: infra/main.tf [hunk 4]
```diff
@@ -23,14 +24,145 @@ resource "aws_security_group" "allow_ssh" {
...
+
+  # When this resources is destroyed, delete the associated key from the file system
+  provisioner "local-exec" {
+    when    = destroy
+    command = "rm ~/.ssh/${self.id}.pem"
+  }
 }
 
 resource "aws_instance" "poc_instance" {
+  depends_on = [
```
Comment (lmilbaum) [inline infra/main.tf:108]: Is it really required?

File: infra/main.tf [hunk 5]
```diff
@@ -23,14 +24,145 @@ resource "aws_security_group" "allow_ssh" {
...
+  }
+}
 
-  tags = var.tags
+resource "aws_secretsmanager_secret" "poc_secret" {
+  name        = var.secret_name
+  description = "secret to store the credentials"
+}
+
+# resource "aws_secretsmanager_secret_rotation" "poc_secret" {
```
Comment (lmilbaum) [inline infra/main.tf]: Remove comments

File: infra/main.tf [hunk 6]
```diff
@@ -23,14 +24,145 @@ resource "aws_security_group" "allow_ssh" {
...
+#   # rotation_lambda_arn = aws_lambda_function.poc_func.arn
+#   rotation_rules {
+#     automatically_after_days = var.rotation_days
+#   }
+# }
+
+resource "aws_secretsmanager_secret_version" "poc_secret" {
+  secret_id     = aws_secretsmanager_secret.poc_secret.id
+  secret_string = jsonencode(var.secret)
+  # Once Secrets Manager has managed to rotate the password,
```
Comment (lmilbaum) [inline infra/main.tf]: Secrets rotation is not in the scope of the POC

**Review by lmilbaum — CHANGES_REQUESTED** [review]

File: infra/variable.tf [hunk 7]
```diff
@@ -80,6 +50,20 @@ variable "profile_name" {
...
 variable "tags" {
   type        = map(string)
-  default     = { "name" : "POC" }
+  default     = { "project" : "POC" }
   description = "Resource tags"
 }
+
+variable "secret_name" {
+  type    = string
+  default = "poc_secret"
```
Comment (lmilbaum) [inline infra/variable.tf]: Currently,  the secret name `poc` is hard coded in the main.py

**Review by lmilbaum — CHANGES_REQUESTED** [review]

File: infra/main.tf (hunk 4 above)
Comment (yerdreic) [reply infra/main.tf:108]: I am not sure actually.

Want me to try removing it?

Comment (lmilbaum) [reply infra/main.tf:108]: Not now

**Review by lmilbaum — APPROVED** [review]
LGTM

**Pull request merged**